lib/ddtrace/contrib/mongodb/patcher.rb

```diff
@@ -0,0 +1,93 @@
+# requirements should be kept minimal as Patcher is a shared requirement.
+
+module Datadog
+  module Contrib
+    # MongoDB module includes classes and functions to instrument MongoDB clients
+    module MongoDB
+      APP = 'mongodb'.freeze
+      SERVICE = 'mongodb'.freeze
+
+      # Patcher adds subscribers to the MongoDB driver so that each command is traced.
+      # Use the `Datadog::Monkey.patch_module(:mongodb)` to activate tracing for
+      # this module.
+      module Patcher
+        @patched = false
+
+        module_function
+
+        def patched?
+          @patched
+        end
+
+        def patch
+          # versions prior to 2.1.0 don't support the Monitoring API
+          if !@patched && (defined?(::Mongo::Monitoring::Global) && \
+                  Gem::Version.new(::Mongo::VERSION) >= Gem::Version.new('2.1.0'))
+            begin
+              require 'ddtrace/pin'
+              require 'ddtrace/ext/net'
+              require 'ddtrace/ext/mongo'
+              require 'ddtrace/ext/app_types'
+              require 'ddtrace/contrib/mongodb/parsers'
+              require 'ddtrace/contrib/mongodb/subscribers'
+
+              patch_mongo_client()
+              add_mongo_monitoring()
+
+              @patched = true
+            rescue StandardError => e
+              Datadog::Tracer.log.error("Unable to apply MongoDB integration: #{e}")
+            end
+          end
+          @patched
+        end
+
+        def add_mongo_monitoring
+          # Subscribe to all COMMAND queries with our subscriber class
+          ::Mongo::Monitoring::Global.subscribe(
+            ::Mongo::Monitoring::COMMAND,
+            Datadog::Contrib::MongoDB::MongoCommandSubscriber.new
+          )
+        end
+
+        def patch_mongo_client
+          ::Mongo::Client.class_eval do
+            alias_method :initialize_without_datadog, :initialize
+            Datadog::Monkey.without_warnings do
+              remove_method :initialize
+            end
+
+            def initialize(*args)
```

Comment:
The patched constructor is not fully compatible with the original one, which accepts a block: http://api.mongodb.com/ruby/current/Mongo/Client.html
You can fix that by doing this:
```rb
def initialize(*args, &blk)
  initialize_without_datadog(*args, &blk)
end
```

Reply:
right!

Reply:
Added a test for that so we're sure there is no regression